### PR TITLE
gh-90300: split --help output into separate options

### LINKED ISSUE #90300: python --help output is too long
BPO | [46142](https://bugs.python.org/issue46142)
--- | :---
Nosy | @warsaw, @vstinner, @merwok, @serhiy-storchaka, @eryksun
PRs | <li>python/cpython#30331</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

Linked PRs:

- #93415

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-12-20.23:43:15.950>
labels = ['interpreter-core', 'type-feature', '3.11']
title = 'python --help output is too long'
updated_at = <Date 2022-01-12.01:45:18.299>
user = 'https://github.com/merwok'
```

bugs.python.org fields:
```python
activity = <Date 2022-01-12.01:45:18.299>
actor = 'eric.araujo'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core']
creation = <Date 2021-12-20.23:43:15.950>
creator = 'eric.araujo'
dependencies = []
files = []
hgrepos = []
issue_num = 46142
keywords = ['patch']
message_count = 15.0
messages = ['408981', '408982', '409449', '409450', '409470', '409483', '409738', '409890', '409928', '410052', '410203', '410207', '410238', '410239', '410378']
nosy_count = 5.0
nosy_names = ['barry', 'vstinner', 'eric.araujo', 'serhiy.storchaka', 'eryksun']
pr_nums = ['30331']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue46142'
versions = ['Python 3.11']
```

</p></details>

### Linked PRs
* gh-93415
* gh-115847
* gh-115853
* gh-116418
* gh-116721
* gh-116724
* gh-116725
* gh-116739
* gh-116756
* gh-116765
* gh-116766
* gh-116779
* gh-116786
* gh-116796
* gh-116797
* gh-116803
* gh-117022
* gh-117034
* gh-117035

### PR DESCRIPTION
**Pull request by merwok**
gh-90300: Make --help output shorter and add new help options.

--help-env, --help-xoptions and --help-all command-line options are
added to complement --help.

### CODE REVIEW

File: Python/initconfig.c [hunk 1]
```diff
@@ -63,8 +63,12 @@ static const char usage_3[] = "\
 -W arg : warning control; arg is action:message:category:module:lineno\n\
          also PYTHONWARNINGS=arg\n\
 -x     : skip first line of source, allowing use of non-Unix forms of #!cmd\n\
--X opt : set implementation-specific option. The following options are available:\n\
-\n\
+-X opt : set implementation-specific option (see details with -X help)\n\
+--check-hash-based-pycs always|default|never:\n\
+         control how Python invalidates hash-based .pyc files\n\
+";
+static const char usage_xoptions[] = "\
```
Comment (merwok) [inline Python/initconfig.c]: For now this string is between usage_3 and usage_4, to minimize the diff.
I think I will move it, and change the formatting as well (don’t need all these leading spaces, may need some blank lines)

File: Python/initconfig.c [hunk 2]
```diff
@@ -2222,6 +2224,11 @@ config_usage(int error, const wchar_t* program)
     }
 }
 
+static void
+config_xoptions_usage()
+{
+    fputs(usage_xoptions, stdout);
+}
```
Comment (merwok) [inline Python/initconfig.c:2276]: Maybe this does not need to be a function.  I started by copying config_usage but ended up with a single line 🙂 
If someone requests that an invalid X option should print help, then I’ll re-add the code to choose stdout / stderr.

File: Python/initconfig.c [hunk 3]
```diff
@@ -99,9 +103,6 @@ static const char usage_3[] = "\
             when the interpreter displays tracebacks.\n\
          -X frozen_modules=[on|off]: whether or not frozen modules should be used.\n\
             The default is \"on\" (or \"off\" if you are running a local build).\n\
-\n\
---check-hash-based-pycs always|default|never:\n\
-    control how Python invalidates hash-based .pyc files\n\
 ";
 static const char usage_4[] = "\
```
Comment (merwok) [inline Python/initconfig.c]: I don’t know why the help text is `split into parts < 512 bytes` (comment on line 31); usage_xoptions (was usage_3) doesn’t seem to be, and it has worked fine.

File: Python/initconfig.c [hunk 4]
```diff
@@ -2232,6 +2239,7 @@ config_parse_cmdline(PyConfig *config, PyWideStringList *warnoptions,
     const PyWideStringList *argv = &config->argv;
     int print_version = 0;
     const wchar_t* program = config->program_name;
+    const wchar_t* xoption_help;
```
Comment (merwok) [inline Python/initconfig.c]: Moved this here to appease a warning.  I hope `const` is correct!

File: Lib/test/test_cmd_line.py [hunk 5]
```diff
@@ -89,12 +97,20 @@ def get_xoptions(*args):
     @unittest.skipIf(interpreter_requires_environment(),
                      'Cannot run -E tests when PYTHON env vars are required.')
     def test_unknown_xoptions(self):
-        rc, out, err = assert_python_failure('-X', 'blech')
+        _, out, err = assert_python_failure('-X', 'blech')
         self.assertIn(b'Unknown value for option -X', err)
-        msg = b'Fatal Python error: Unknown value for option -X'
+        msg = b'Fatal Python error: Unknown value for option -X (see -X help)'
```
Comment (merwok) [inline Lib/test/test_cmd_line.py]: Lines 101 and 120 seem redundant but I kept both.

File: Misc/python.man [hunk 6]
```diff
@@ -66,10 +66,10 @@ python \- an interpreted, interactive, object-oriented programming language
 .B \-x
 ]
 [
-[
 .B \-X
 .I option
 ]
+[
```
Comment (merwok) [inline Misc/python.man]: Drive-by fix, should backport

File: Misc/python.man [hunk 7]
```diff
@@ -81,6 +81,16 @@ python \- an interpreted, interactive, object-oriented programming language
...
+.br
+       [
+.B \--help
+]
+[
+.B \--help-env
+]
+[
+.B \-X help
+]
```
Comment (merwok) [inline Misc/python.man:99]: I thought showing all help options on their own line could be nice.

File: Python/initconfig.c [hunk 8]
```diff
@@ -2274,8 +2295,11 @@ config_parse_cmdline(PyConfig *config, PyWideStringList *warnoptions,
 
         switch (c) {
         case 0:
-            // Handle long option.
-            assert(longindex == 0); // Only one long option now.
+          // Handle long options.
```
Comment (merwok) [inline Python/initconfig.c]: 2 spaces temporarily to minimize the diff for reviews — will move to 4

File: Python/initconfig.c [hunk 9]
```diff
@@ -2274,8 +2295,11 @@ config_parse_cmdline(PyConfig *config, PyWideStringList *warnoptions,
...
+          assert(longindex < 2);  // Only two long options for now.
+          switch(longindex) {
+          case 0:
```
Comment (merwok) [inline Python/initconfig.c]: I think I could actually avoid this nested switch by setting a different `val` in the long option struct I added, let me try that

File: Python/initconfig.c [hunk 10]
```diff
@@ -63,8 +63,20 @@ static const char usage_3[] = "\
 -W arg : warning control; arg is action:message:category:module:lineno\n\
          also PYTHONWARNINGS=arg\n\
 -x     : skip first line of source, allowing use of non-Unix forms of #!cmd\n\
--X opt : set implementation-specific option. The following options are available:\n\
-\n\
+-X opt : set implementation-specific option (see details with -X help)\n\
+--help-env : print help about other environment variables\n\
```
Comment (serhiy-storchaka) [inline Python/initconfig.c]: Why other? Why not all?

File: Python/initconfig.c [hunk 11]
```diff
@@ -2293,6 +2314,12 @@ config_parse_cmdline(PyConfig *config, PyWideStringList *warnoptions,
             }
             break;
 
+        case 1:
+            // help-env
+            config_envvars_usage();
+            return _PyStatus_EXIT(0);
+            break;
```
Comment (serhiy-storchaka) [inline Python/initconfig.c]: `break` is redundant.

File: Doc/using/cmdline.rst [hunk 12]
```diff
@@ -183,15 +183,21 @@ automatically enabled, if available on your platform (see
...
                -h
                --help
 
-   Print a short description of all command line options.
+   Print a short description of all command line options and related
+   environment variables.
+
+.. cmdoption:: --help-env
 
+   Print a short description of additional environment variables.
```
Comment (warsaw) [inline Doc/using/cmdline.rst]: Do you think this text is confusing?  Will the reader understand the difference between "related" environment variables and "additional" environment variables?  Maybe this text should mirror what `-h` also says:

```
% ./python.exe -h
usage: (null) [option] ... [-c cmd | -m mod | file | -] [arg] ...
Options and arguments (and corresponding environment variables):
```

(i.e. s/related/corresponding/)?

File: Doc/using/cmdline.rst [hunk 13]
```diff
@@ -447,6 +458,7 @@ Miscellaneous options
    Reserved for various implementation-specific options.  CPython currently
    defines the following possible values:
 
+   * ``-X help`` to print a short description of all X options;
```
Comment (warsaw) [inline Doc/using/cmdline.rst]: Perhaps: "all -X options"?

File: Lib/test/test_cmd_line.py (hunk 5 above)
Comment (warsaw) [reply Lib/test/test_cmd_line.py]: It might be nice if this could print all the offending `-X` options, but that seems difficult given the interface to `_Py_check_xoptions()` and `config_read()`.

File: Python/initconfig.c (hunk 10 above)
Comment (warsaw) [inline Python/initconfig.c]: "all" instead of "other"?

File: Python/initconfig.c [hunk 14]
```diff
@@ -63,8 +63,20 @@ static const char usage_3[] = "\
...
+";
+static const char usage_4[] = "\
+file   : program read from script file\n\
+-      : program read from stdin (default; interactive mode if a tty)\n\
+arg ...: arguments passed to program in sys.argv[1:]\n\
+";
+
+static const char usage_xoptions[] = "\
+The following implementation-specific options are available:\n\
+         -X help: show this help\n\
```
Comment (warsaw) [inline Python/initconfig.c]: ...and exit, right?

File: Python/initconfig.c (hunk 2 above)
Comment (warsaw) [reply Python/initconfig.c:2276]: Invalid `-X` option output could be improved, although I'm just not sure if it's worth additional complexity:

```
% ./python.exe -X foo
Fatal Python error: Unknown value for option -X (see -X help)
Python runtime state: preinitialized
```

File: Python/initconfig.c (hunk 4 above)
Comment (warsaw) [reply Python/initconfig.c]: Compiler Knows Best 😄

**Review by warsaw — CHANGES_REQUESTED** [review]

**Comment by warsaw** [comment]
Thanks for working on this!  I left some comments and suggestions.

File: Python/initconfig.c (hunk 2 above)
Comment (merwok) [reply Python/initconfig.c:2276]: Is your example output here what you would like?  Because the PR does add `(see -X help)` 🙂

File: Lib/test/test_cmd_line.py (hunk 5 above)
Comment (merwok) [reply Lib/test/test_cmd_line.py]: Hmm I think something can be shown, either the unknown value or the raw argv.  I’ll have a look.

File: Doc/using/cmdline.rst (hunk 13 above)
Comment (merwok) [reply Doc/using/cmdline.rst]:
```suggestion
   * ``-X help`` to print a short description of all ``-X`` options;
```

File: Doc/using/cmdline.rst [hunk 15]
```diff
@@ -183,15 +183,21 @@ automatically enabled, if available on your platform (see
...
+
 Generic options
 ~~~~~~~~~~~~~~~
 
 .. cmdoption:: -?
                -h
                --help
 
-   Print a short description of all command line options.
+   Print a short description of all command line options and related
```
Comment (merwok) [inline Doc/using/cmdline.rst]:
```suggestion
   Print a short description of all command line options and corresponding
```

File: Doc/using/cmdline.rst (hunk 12 above)
Comment (merwok) [reply Doc/using/cmdline.rst]: Yeah, the various help texts need improvement.
+1 for the change to `related`, and let’s hope that `additional` here will make sense.
Or I could go the long way:

>  Print a short description of environment variables that are observed by Python and do not correspond to a command-line option.

File: Python/initconfig.c (hunk 10 above)
Comment (merwok) [reply Python/initconfig.c]: Other means: other than the ones that you can read about just above in this help output

File: Python/initconfig.c (hunk 14 above)
Comment (merwok) [reply Python/initconfig.c]:
```suggestion
         -X help: show this help and exit\n\
```

File: Lib/test/test_cmd_line.py (hunk 5 above)
Comment (merwok) [reply Lib/test/test_cmd_line.py:102]: Can’t show unknown value, _Py_check_xoptions returns a known string or NULL when invalid.

File: Python/initconfig.c (hunk 2 above)
Comment (warsaw) [reply Python/initconfig.c:2276]: It might be nicer to mimic what other invalid options print, e.g.:

```
% ./python.exe -X foo
Unknown option: -X foo
usage: (null) [option] ... [-c cmd | -m mod | file | -] [arg] ...
Try `python -h' for more information.
```

although maybe `Try python -X help for more information`?

File: Doc/using/cmdline.rst (hunk 12 above)
Comment (warsaw) [reply Doc/using/cmdline.rst]: Works for me, thanks.

File: Python/initconfig.c (hunk 10 above)
Comment (warsaw) [reply Python/initconfig.c]: Oh, I see.  Yeah, that's a little weird that `-X help` doesn't actually print *all* the `-X` options.  How hard would that be?

File: Lib/test/test_cmd_line.py (hunk 5 above)
Comment (merwok) [reply Lib/test/test_cmd_line.py]: Turns out the get xoptions method returns NULL for valid options, otherwise the invalid value!
I just need help to format the error string: I don’t know how to format the wchar_t string `option` into a const char error message.

```c
    /* -X options */
    const wchar_t* option = _Py_check_xoptions(&config->xoptions, known_xoptions);
    char errmsg[100];
    if (option != NULL) {
        // XXX gibberish printed with %S, %.50S, %ls + type conversion errors (same with snprintf)
        sprintf(errmsg, "Unknown value for option -X: %ls (see -X help)", option);
        return PyStatus_Error(errmsg);
    }
```

File: Python/initconfig.c (hunk 10 above)
Comment (merwok) [reply Python/initconfig.c]: `-X help` is fine 😸   Here we are saying that `--help-env` does not print all the environment variables.
It could be done, at the little price of duplicating the info in the `--help` output.

For example, we would have both these lines:
```c
usage_2 = \
"--enable-flufl    : enable flufl to frobnicate the gruzles (also PYTHONENABLEFLUFL=1)\n"

usage_envvars_1 = \
"PYTHONENABLEFLUFL=1  : enable flufl to frobnicate the gruzles (also --enable-flufl)\n"
```

Does that seem good?

File: Python/initconfig.c (hunk 2 above)
Comment (merwok) [reply Python/initconfig.c:2276]: Because it’s a fatal error, I am unsure about changing the string format.
Keeping two lines with `Fatal Python error: details details` and `Python runtime state: info` seems better.

Comment (warsaw) [reply Python/initconfig.c:2276]: Why would `-X foo` be a fatal error but not `--foo`?  IOW, it feels inconsistent, but maybe this isn't important enough to increase complexity.

File: Lib/test/test_cmd_line.py (hunk 5 above)
Comment (warsaw) [reply Lib/test/test_cmd_line.py]: > I just need help to format the error string: I don’t know how to format the wchar_t string option into a const char error message.

Wouldn't it be nicer if this could all be done in Python? 😄 

If it's difficult, don't worry about.

File: Python/initconfig.c (hunk 10 above)
Comment (warsaw) [reply Python/initconfig.c]: Oops, yep, I typoed that!

> Does that seem good?

I like it!  You get the fully details either way.  And I say that as an experienced gruzle frobnicator!

File: Python/initconfig.c (hunk 2 above)
Comment (merwok) [reply Python/initconfig.c:2276]: Probably because some X options change fundamental behaviour, so they are enabled during early startup and cause fatal errors.

File: Lib/test/test_cmd_line.py (hunk 5 above)
Comment (merwok) [reply Lib/test/test_cmd_line.py]: It’s probably simple string conversion and formatting to a C programmer, but I am only playing one on TV here.

File: Python/initconfig.c (hunk 2 above)
Comment (warsaw) [reply Python/initconfig.c:2276]: > Probably because some X options change fundamental behaviour, so they are enabled during early startup and cause fatal errors.

Yep, in which case, let's just leave it as you have it.

File: Python/initconfig.c (hunk 11 above)
Comment (ericsnowcurrently) [inline Python/initconfig.c]: not needed?

**Review by vstinner — COMMENTED** [review]
``-X dev`` sounds inconsistent with ``--help`` and ``--help-env``. Why not ``--help-xoptions``?

Note: I like replacing ``git command ...`` with ``git help command`` and that it just works :-) For example, if ``git bisect ...`` fails, I get the help by running ``git help bisect``. But Python CLI is very different, it doesn't use sub-commands.

**Comment by serhiy-storchaka** [comment]
> `-X dev` sounds inconsistent with `--help` and `--help-env`. Why not `--help-xoptions`?

See https://mail.python.org/archives/list/python-dev@python.org/thread/QUUBM7DGSXYWBOLZNWOSCQUDALWJIYZF/ .

Guido proposed it and everybody liked it.

File: Python/initconfig.c (hunk 10 above)
Comment (merwok) [reply Python/initconfig.c]: See reply to same question from Barry

File: Python/initconfig.c (hunk 11 above)
Comment (merwok) [reply Python/initconfig.c]: Thanks, Eric Snow already noted that and I fixed it locally.

**Comment by vstinner** [comment]
> Guido proposed it and everybody liked it.

Well, I dislike `python -X help`, I would prefer `python --help-xoptions` :-) But the majority wins.

File: Doc/using/cmdline.rst [hunk 16]
```diff
@@ -183,15 +183,21 @@ automatically enabled, if available on your platform (see
...
                -h
                --help
 
-   Print a short description of all command line options.
+   Print a short description of all command line options and corresponding
+   environment variables.
+
+.. cmdoption:: --help-env
 
+   Print a short description of Python-specific environment variables.
```
Comment (merwok) [inline Doc/using/cmdline.rst]: I didn’t say «all env vars» because CPython also observes PATH, LANG, etc.

File: Python/initconfig.c [hunk 17]
```diff
@@ -63,8 +64,21 @@ static const char usage_3[] = "\
...
 -x     : skip first line of source, allowing use of non-Unix forms of #!cmd\n\
--X opt : set implementation-specific option. The following options are available:\n\
-\n\
+-X opt : set implementation-specific option (see details with -X help)\n\
+--help-env : print help about Python-specific environment variables\n\
+--check-hash-based-pycs always|default|never :\n\
+         control how Python invalidates hash-based .pyc files\n\
+";
+static const char usage_4[] = "\
+Arguments:\n\
```
Comment (merwok) [inline Python/initconfig.c:70]: Moved from line 34; I thought these params deserved their group.  Could even move them before options.

File: Python/initconfig.c [hunk 18]
```diff
@@ -141,6 +153,17 @@ static const char usage_6[] =
...
+static const char usage_envvars5[] = "\
+These variables have equivalent command-line parameters (see --help for details):\n\
+PYTHONDEBUG             : enable parser debug mode (-d)\n\
+PYTHONDONTWRITEBYTECODE : don't write .pyc files (-B)\n\
+PYTHONINSPECT           : inspect interactively after running script (-i)\n\
+PYTHONNOUSERSITE        : disable user site directory (-s)\n\
+PYTHONOPTIMIZE          : enable level 1 optimizations (-O)\n\
+PYTHONUNBUFFERED        : disable stdout/stderr buffering (-u)\n\
+PYTHONVERBOSE           : trace import statements (-v)\n\
+PYTHONWARNINGS=arg      : warning control (-W arg)\n\
```
Comment (merwok) [inline Python/initconfig.c]: @warsaw in the end, I wrote a short help summary for each envvar rather than copying the full lines.
Didn’t want to create the burden of writing the same things twice in the same file, with slightly different formatting.

Of course I am ready to change that if reviewers think full details should be here.  There are only 8 envvars, and the text is already duplicated in source, man page and rst docs.

File: Lib/test/test_cmd_line.py [hunk 19]
```diff
@@ -37,6 +37,20 @@ def verify_valid_flag(self, cmd_line):
...
+        self.verify_valid_flag('--help')
+
+    def test_help_env(self):
+        self.verify_valid_flag('--help-env')
+
+    def test_help_xoptions(self):
+        self.verify_valid_flag('--help-xoptions')
+
+    def test_help_all(self):
+        self.verify_valid_flag('--help-all')
```
Comment (merwok) [inline Lib/test/test_cmd_line.py]: The three new options could get a better test, say checking the first line for a well-known string (like help_usage on line 61 below)

File: Python/initconfig.c [hunk 20]
```diff
@@ -141,6 +151,16 @@ static const char usage_6[] =
 "   and end column offset) to every instruction in code objects. This is useful \n"
 "   when smaller cothe de objects and pyc files are desired as well as suppressing the \n"
 "   extra visual location indicators when the interpreter displays tracebacks.\n";
```
Comment (arhadthedev) [inline Python/initconfig.c]:
```suggestion
"   extra visual location indicators when the interpreter displays tracebacks.";
```
(since other strings affected by  b9a0b19e3 *fputs→puts* transition have their `\n`s stripped)

File: Python/initconfig.c [hunk 21]
```diff
@@ -2216,12 +2236,36 @@ config_usage(int error, const wchar_t* program)
...
+config_xoptions_usage()
+{
+    puts(usage_xoptions);
+}
+
+static void
+config_complete_usage(const wchar_t* program)
+{
+   config_usage(0, program);
+   puts("\n");
```
Comment (arhadthedev) [inline Python/initconfig.c:2282]: Will `putchar` be better for a single letter output? I mean, semantically, not for individual bytes and nanoseconds.

```suggestion
   putchar('\n');
```

File: Python/initconfig.c [hunk 22]
```diff
@@ -2216,12 +2236,36 @@ config_usage(int error, const wchar_t* program)
...
+   config_envvars_usage();
+   puts("\n");
```
Comment (arhadthedev) [inline Python/initconfig.c:2284]: Ditto.

```suggestion
   putchar('\n');
```

**Review by arhadthedev — CHANGES_REQUESTED** [review]

File: Python/initconfig.c (hunk 20 above)
Comment (merwok) [reply Python/initconfig.c]: it’s only strings that are last of their section, newlines are still needed between items

File: Python/initconfig.c (hunk 21 above)
Comment (merwok) [reply Python/initconfig.c:2282]: The intent here is «put things between sections», not «put one char specifically», so no :)

**Comment by merwok** [comment]
I would love to get this into 3.11, but there are some questions around error reporting and output formatting (see the unresolved discussions in this page)

**Comment by serhiy-storchaka** [comment]
How can I help @merwok?

File: Lib/test/test_cmd_line.py (hunk 5 above)
Comment (merwok) [reply Lib/test/test_cmd_line.py]: @serhiy-storchaka my issue was with C string handling here!  thanks for the help

Comment (serhiy-storchaka) [reply Lib/test/test_cmd_line.py]: `%ls` should work (and it is already used in other places. But if there are any issues with this, you can just omit outputting the option.

Comment (merwok) [reply Lib/test/test_cmd_line.py]: I have seen that, and have tried it (see my message `Turns out etc`), but I couldn’t figure out the right combination of wchar_t, `L""` string and sprintf codes.

I will abandon printing the bad option, but anyway the window for 3.11 is missed.

Comment (serhiy-storchaka) [reply Lib/test/test_cmd_line.py]: It is a nice enhancement and I hope to see it in 3.11. @pablogsal

Comment (pablogsal) [reply Lib/test/test_cmd_line.py]: I would be happy to backport this. Please, add me as a reviewer in the 3.11 backport

**Comment by merwok** [comment]
The outstanding items are:
- [ ] positive reaction to the choice of `--help-xoptions` (chosen for symmetry, discoverability and ease of implementation over the idea of `-X help`)
- [ ] minimal or expanded tests https://github.com/python/cpython/pull/30331/files#r782650181
- [ ] positive or negative reaction to these remarks
    - [ ] https://github.com/python/cpython/pull/30331/files#r777264087
    - [ ] https://github.com/python/cpython/pull/30331/files#r780566821
    - [ ] https://github.com/python/cpython/pull/30331/files#r777169481

File: Lib/test/test_cmd_line.py [hunk 23]
```diff
@@ -28,17 +28,33 @@ def _kill_python_and_exit_code(p):
...
 
+
 class CmdLineTest(unittest.TestCase):
     def test_directories(self):
         assert_python_failure('.')
         assert_python_failure('< .')
 
     def verify_valid_flag(self, cmd_line):
         rc, out, err = assert_python_ok(*cmd_line)
+        self.assertEqual(rc, 0)
```
Comment (serhiy-storchaka) [inline Lib/test/test_cmd_line.py]: It is already checked in `assert_python_ok()`.

File: Python/initconfig.c [hunk 24]
```diff
@@ -39,6 +40,9 @@ Options and arguments (and corresponding environment variables):\n\
          debug builds); also PYTHONDEBUG=x\n\
 -E     : ignore PYTHON* environment variables (such as PYTHONPATH)\n\
 -h     : print this help message and exit (also --help)\n\
+--help-env      : print help about Python-specific environment variables and exit\n\
```
Comment (serhiy-storchaka) [inline Python/initconfig.c]: Line longer than 79-columns. It was one of reasons of rewriting the help output.

File: Python/initconfig.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Python/initconfig.c]: Yes, please remove leading spaces and rewrap all lines.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM, but the final edition should reformat a lot of text, it will increase the diff. Or maybe left it to a separate PR?

File: Python/initconfig.c (hunk 24 above)
Comment (merwok) [reply Python/initconfig.c]: Is the concern here that the source file should fit in 79 columns?  Existing lines already break that.

Or is it that the help output should fit in 79 columns?  What do you think of `print help about Python environment variables and exit`

**Comment by merwok** [comment]
I’m a bit afraid to launch this discussion here, but I have long heard conflicting rules about formatting!

On one hand, we avoid PRs that only do superficial aesthetic changes.  Guido has this word «holistic approach» that I understand to mean, do these small changes when you are already working in that code for another reason.

But, we also want to make things easy to review, and avoid mixing true changes with reformattings.  So how does it work?!

(We used to have a workflow based on pushing, so this seeming paradox was solved by making a separate commit for review, included in the same push.)

File: Python/initconfig.c (hunk 17 above)
Comment (merwok) [reply Python/initconfig.c:70]: @serhiy-storchaka what do you think about moving this arguments section before options?

File: Lib/test/test_cmd_line.py (hunk 19 above)
Comment (serhiy-storchaka) [reply Lib/test/test_cmd_line.py]: Let it not slow down you. If you add additional tests -- good (I would either add an optional parameter in `verify_valid_flag` for substrings which should occur or should not occur in the output, or make it returning the output string). But we can live without this and can add more tests later, for these and for other options.

File: Python/initconfig.c (hunk 17 above)
Comment (serhiy-storchaka) [reply Python/initconfig.c:70]: It is up to you.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
The output should not be wider than 80 (or 79) columns. Currently the limit is exceeded for -X options and envvars.

File: Python/initconfig.c (hunk 17 above)
Comment (merwok) [reply Python/initconfig.c:70]: Decided against it, to have initconfig.c stay close to the man page, and respect the classic posix rule of options before arguments.

File: Lib/test/test_cmd_line.py (hunk 19 above)
Comment (merwok) [reply Lib/test/test_cmd_line.py]: Adding parameters to `verify_valid_flag` doesn’t work (too many options to support checks for contains / matches / does not contain / in stdout / in stderr), but returning output is nice.

File: Python/initconfig.c [hunk 25]
```diff
@@ -28,17 +28,23 @@
...
+         control how Python invalidates hash-based .pyc files\n\
 -d     : turn on parser debugging output (for experts only, only works on\n\
          debug builds); also PYTHONDEBUG=x\n\
 -E     : ignore PYTHON* environment variables (such as PYTHONPATH)\n\
 -h     : print this help message and exit (also --help)\n\
+--help-env      : print help about Python environment variables and exit\n\
+--help-xoptions : print help about implementation-specific -X options and exit\n\
+--help-all      : print complete help information and exit\n\
 ";
 static const char usage_2[] = "\
```
Comment (ericsnowcurrently) [inline Python/initconfig.c]: Would it be a problem to fold `usage_2` and `usage_3` into `usage_1`?  I'm guessing they're artifacts of when older compilers had limits on string literal size.

Comment (merwok) [reply Python/initconfig.c]: Yes there is a comment about usage being `split into parts < 512 bytes` but usage_xoptions doesn’t follow that and things do build.

File: Python/initconfig.c [hunk 26]
```diff
@@ -2085,7 +2106,7 @@ config_read(PyConfig *config, int compute_path_config)
     /* -X options */
     const wchar_t* option = _Py_check_xoptions(&config->xoptions, known_xoptions);
     if (option != NULL) {
-        return PyStatus_Error("Unknown value for option -X");
+        return PyStatus_Error("Unknown value for option -X (see --help-xoptions)");
     }
```
Comment (merwok) [inline Python/initconfig.c:2110]: Turns out that showing the invalid value in this error message is not trivial: real C programmers had issue with it too before https://github.com/python/cpython/pull/28823

File: Python/initconfig.c [hunk 27]
```diff
@@ -64,64 +61,80 @@ static const char usage_3[] = "\
...
-             given directory instead of to the code tree\n\
-         -X warn_default_encoding: enable opt-in EncodingWarning for 'encoding=None'\n\
-         -X no_debug_ranges: disable the inclusion of the tables mapping extra location \n\
-            information (end line, start column offset and end column offset) to every \n\
-            instruction in code objects. This is useful when smaller code objects and pyc \n\
-            files are desired as well as suppressing the extra visual location indicators \n\
-            when the interpreter displays tracebacks.\n\
-         -X frozen_modules=[on|off]: whether or not frozen modules should be used.\n\
-            The default is \"on\" (or \"off\" if you are running a local build).\n\
+-X opt : set implementation-specific option\n\
```
Comment (AA-Turner) [inline Python/initconfig.c:64]: `--help-xoptions` is listed a few items below, but if scanning the list it might be helpful to explicitly mention it here. (Sorry if it's been suggested before & declined!).

```suggestion
-X opt : set implementation-specific option (see --help-xoptions)\n\
```

Comment (merwok) [reply Python/initconfig.c:64]: I’ve added it and removed it.  In the current version, I think it is not needed, as it is a few lines later and in the man page and shown in error message for unknown X option, but have no problem adding it if someone adds 👍🏽 to your comment 🙂

**Comment by merwok** [comment]
This is ready and complete for me.  I did not change all lines to be <80 columns like Serhiy wanted, but the upside is that the diff is very reasonable.  I would like to merge this week-end, or sooner if there are approving reviews, then create the backport PR.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by ericsnowcurrently** [comment]
🎉 Thanks for seeing this through, @merwok!  It's a nice improvement. 😄
